### DISCUSSION TITLE
Generate form with Json Schema

### DISCUSSION THREAD

**Discussion by ChristoRibeiro** [react-hook-form/react-hook-form#5498] [Q&A] [answered]
Hi,

Is there any libraries or built-in component to use react-hook-form with [Json-schema](https://json-schema.org/)?

Working easily with Json-schema could boost productivity for developers and open opportunities for business oriented people (see [Formium](https://formium.io), headless form platform using the formik library). 

I found the library [react-hook-form-jsonschema](https://github.com/vtex/react-hook-form-jsonschema) but it looks not active anymore.

**Comment by leapful** [comment, marked as answer]
@ChristoRibeiro 

You can use any JSON-Schema validator library to perform custom validation with the option "resolver" of "useForm" hook.

For example with "ajv" (https://github.com/ajv-validator/ajv):

```
import Ajv from 'ajv';

const ajv = new Ajv();

const schema = {
   // JSON schema 
}

useForm({
  resolver: (values) => {
      const validator = ajv.compile(schema);

      const valid = validator(values);

     if (!valid) {
           // Build "errors" object from "validator.errors"
          return {  values: {}, errors }
      }
      
      return { values, errors: {} };
  }
});

```

**Reply by ChristoRibeiro** [reply]
Didn't know about that. Thanks! 

My concern is more about to generate a form based on a json-schema.

**Reply by leapful** [reply]
@ChristoRibeiro 

If so, "react-hook-form" can be used with "react-jsonschema-form" to build form fields with JSON schema.

By disabling built-in validation and overriding FieldElement of "react-jsonschema-form", "react-hook-form" can be applied.

```
import Form from 'react-jsonschema-form';
import { useForm, useFormContext, FormProvider } from 'react-hook-form';

const HookFormStringField = () => {
   const { register } = useFormContext();
   return <input {...register('field')}  />
}

const field = {
   StringField: HookFormStringField
}

const App = () => {
   return <FormProvider>
<Form noValidate fields={fields} >
   <button type="submit">Submit</button>
</Form>
</FormProvider>
}

```

CSB sample: https://codesandbox.io/s/react-json-schema-form-forked-p30jb?file=/src/index.js

**Reply by ChristoRibeiro** [reply]
Nice I can use react-hook-form and react-jsonschema-form together! 

Thank you for you help @leapful 👍

**Reply by zyasir** [reply]
Very helpful, I was also looking to use react-jsonchema-form with react-hook-ofrm. @ChristoRibeiro do you have any results you can share about your experience with this combinatoin?
